src/Http/WebUtilities/src/FormPipeReader.cs

```diff
@@ -0,0 +1,345 @@
+                    || !sequenceReader.IsNext(equalsEncoded, true))
+                {
+                    break;
+                }
+
+                if (!sequenceReader.TryReadTo(out ReadOnlySequence<byte> value, andEncoded, false)
+                    || !sequenceReader.IsNext(andEncoded, true))
+                {
+                    if (!isFinalBlock)
+                    {
+                        break;
+                    }
+
+                    value = buffer.Slice(sequenceReader.Position);
+
+                    sequenceReader.Advance(value.Length);
+                }
+
+                // Need to call ToArray if the key/value spans multiple segments 
+                var decodedKey = GetDecodedString((key.IsSingleSegment ? key.First : key.ToArray()).Span, encoding);
+                var decodedValue = GetDecodedString((value.IsSingleSegment ? value.First : value.ToArray()).Span, encoding);
+
+                AppendAndVerify(ref accumulator, keyLengthLimit, valueLengthLimit, valueCountLimit, decodedKey, decodedValue);
+
+                consumed = sequenceReader.Position;
+            }
+
+            buffer = buffer.Slice(consumed);
+        }
+
+        // Check that key/value constraints are met and appends value to accumulator.
+        private static KeyValueAccumulator AppendAndVerify(ref KeyValueAccumulator accumulator, int keyLengthLimit, int valueLengthLimit, int valueCountLimit, string decodedKey, string decodedValue)
+        {
+            if (decodedKey.Length > keyLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {keyLengthLimit} exceeded.");
+            }
+
+            if (decodedValue.Length > valueLengthLimit)
+            {
+                throw new InvalidDataException($"Form key or value length limit {valueLengthLimit} exceeded.");
+            }
+
+            accumulator.Append(decodedKey, decodedValue);
+
+            if (accumulator.ValueCount > valueCountLimit)
+            {
+                throw new InvalidDataException($"Form value count limit {valueCountLimit} exceeded.");
+            }
+
+            return accumulator;
+        }
+
+        private static string GetDecodedString(ReadOnlySpan<byte> readOnlySpan, Encoding encoding)
+        {
+            if (readOnlySpan.Length == 0)
+            {
+                return "";
+            }
+            else if (encoding == Encoding.UTF8 || encoding == Encoding.ASCII)
+            {
+                // UrlDecoder only works on UTF8 (and implicitly ASCII)
+
+                // We need to create a Span from a ReadOnlySpan. This cast is safe because the memory is still held by the pipe
+                // We will also create a string from it by the end of the function.
+                var span = MemoryMarshal.CreateSpan(ref Unsafe.AsRef(readOnlySpan[0]), readOnlySpan.Length);
+                var bytes = UrlDecoder.DecodeInPlace(span);
```

Comment:
We'll need to think about how DecodeInPlace affects buffered/seekable data. This span is likely a temp buffer and not the real request body buffer, but keep that in mind.